Fix calculating load_factor for HASHED dictionaries with SHARDS

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `load_factor`/`element_count` for `HASHED`/`SPARSE_HASHED` dictionaries with `SHARDS` or multiple attributes

Follow-up for: #40003
Fixes: #5440
Cc: @kitaisreal

azat committed:
Fix load_factor for HASHED/SPARSE_HASHED dictionaries with SHARDS

Previously, bucket_count was set only for the one shard, and hence
load_factor was > 1.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

vdimir commented on src/Dictionaries/HashedDictionary.cpp:
```diff
@@ -1023,6 +1024,8 @@ void HashedDictionary<dictionary_key_type, sparse, sharded>::calculateBytesAlloc
     for (const auto & arena : string_arenas)
         bytes_allocated += arena->allocatedBytes();
+
+    bucket_count = new_bucket_count;
```
Why don't we reset `bytes_allocated` in the same way?

azat replied:
Because `bytes_allocated` should be a sum across all attributes, while `bucket_count` should not (since it is used to calculate the `load_factor` as `element_count / bucket_count`)

But your question made me look into this deeply, and I found that `element_count` was incorrectly calculated for dictionaries with multiple attributes, so I've rebased the patch set, add some comments and update the test.

azat committed:
Fix element_count for HASHED/SPARSE_HASHED with multiple attributes

Previosly element_count was multplied by the number of attributes.

Fixes: #5440
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Improve 02730_dictionary_hashed_load_factor test

- add attributes & names into output
- rename to include element_count into the name too

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #49319 from azat/ClickHouse:dict/fix-hashed-load_factor -->
This is an automated comment for commit 5fbf601764d78f886a9a59458712cdd19c3fc5a7 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49319/5fbf601764d78f886a9a59458712cdd19c3fc5a7/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

azat commented:
>Upgrade check (debug) — Possible deadlock on shutdown (see gdb.log)

Just slow S3?

```
2023-05-01 18:22:35 Thread 255 (Thread 0x7f98fbde1700 (LWP 191316)):
2023-05-01 18:22:35 #0  0x00000000324d44ac in tsd_state_get (tsd=0x7f98fbdd4ac0) at /build/contrib/jemalloc/include/jemalloc/internal/tsd.h:269
2023-05-01 18:22:35 #1  0x00000000324db395 in tsd_reentrancy_levelp_get (tsd=0x7f98fbdd4ac0) at /build/contrib/jemalloc/include/jemalloc/internal/tsd.h:344
2023-05-01 18:22:35 #2  0x00000000324c8ef5 in tsd_reentrancy_level_get (tsd=0x7f98fbdd4ac0) at /build/contrib/jemalloc/include/jemalloc/internal/tsd.h:373
2023-05-01 18:22:35 #3  0x00000000324c32e0 in check_entry_exit_locking (tsdn=0x7f98fbdd4ac0) at /build/contrib/jemalloc/src/jemalloc.c:687
2023-05-01 18:22:35 #4  0x00000000324c35a0 in ifree (tsd=0x7f98fbdd4ac0, ptr=0x7f996e6dea00, tcache=0x7f98fbdd4e18, slow_path=true) at /build/contrib/jemalloc/src/jemalloc.c:2868
2023-05-01 18:22:35 #5  0x00000000324c3255 in free_default (ptr=0x7f996e6dea00) at /build/contrib/jemalloc/src/jemalloc.c:3021
2023-05-01 18:22:35 #6  0x00000000324c3905 in free (ptr=0x7f996e6dea00) at /build/contrib/jemalloc/src/jemalloc.c:3162
2023-05-01 18:22:35 #7  0x0000000032986e4e in Aws::Free (memoryPtr=0x7f996e6dea00) at /build/contrib/aws/aws-cpp-sdk-core/source/utils/memory/AWSMemory.cpp:97
2023-05-01 18:22:35 #8  0x0000000032986f99 in Aws::MemRelease (allocator=0x37a3cf78 <Aws::get_aws_allocator()::wrapper>, ptr=0x7f996e6dea00) at /build/contrib/aws/aws-cpp-sdk-core/source/utils/memory/AWSMemory.cpp:110
2023-05-01 18:22:35 #9  0x0000000032c0741b in aws_mem_release (allocator=0x37a3cf78 <Aws::get_aws_allocator()::wrapper>, ptr=0x7f996e6dea00) at /build/contrib/aws-c-common/source/allocator.c:213
2023-05-01 18:22:35 #10 0x0000000032c3fd65 in aws_hash_table_clean_up (map=0x7f98b3a91b30) at /build/contrib/aws-c-common/source/hash_table.c:261
2023-05-01 18:22:35 #11 0x0000000032da8fd1 in s_endpoints_request_context_destroy (data=0x7f98b3a91b10) at /build/contrib/aws-c-sdkutils/source/endpoints_rule_engine.c:652
2023-05-01 18:22:35 #12 0x0000000032c3f3cd in aws_ref_count_release (ref_count=0x7f98b3a91b18) at /build/contrib/aws-c-common/source/ref_count.c:29
2023-05-01 18:22:35 #13 0x0000000032da9084 in aws_endpoints_request_context_release (request_context=0x7f98b3a91b10) at /build/contrib/aws-c-sdkutils/source/endpoints_rule_engine.c:697
2023-05-01 18:22:35 #14 0x0000000032d296bd in Aws::Crt::Endpoints::RequestContext::~RequestContext (this=0x7f98fbdd01a8) at /build/contrib/aws-crt-cpp/source/endpoints/RuleEngine.cpp:25
2023-05-01 18:22:35 #15 0x000000003292d19f in Aws::Endpoint::ResolveEndpointDefaultImpl (ruleEngine=..., builtInParameters=..., clientContextParameters=..., endpointParameters=...) at /build/contrib/aws/aws-cpp-sdk-core/source/endpoint/DefaultEndpointProvider.cpp:225
2023-05-01 18:22:35 #16 0x0000000026b4ea58 in Aws::Endpoint::DefaultEndpointProvider<Aws::S3::S3ClientConfiguration, Aws::S3::Endpoint::S3BuiltInParameters, Aws::S3::Endpoint::S3ClientContextParameters>::ResolveEndpoint (this=0x7f998736efd8, endpointParameters=...) at /build/contrib/aws/aws-cpp-sdk-core/include/aws/core/endpoint/DefaultEndpointProvider.h:76
2023-05-01 18:22:35 #17 0x00000000329c08c5 in Aws::S3::S3Client::DeleteObjects (this=0x7f9987305800, request=...) at /build/contrib/aws/aws-cpp-sdk-s3/source/S3Client.cpp:969
2023-05-01 18:22:35 #18 0x000000002989c30b in DB::S3::Client::DeleteObjects(DB::S3::ExtendedRequest<Aws::S3::Model::DeleteObjectsRequest> const&) const::$_11::operator()(Aws::S3::Model::DeleteObjectsRequest const&) const (this=0x7f98fbdd1a80, req=...) at /build/src/IO/S3/Client.cpp:317
2023-05-01 18:22:35 #19 0x0000000029897216 in DB::S3::Client::doRequest<DB::S3::ExtendedRequest<Aws::S3::Model::DeleteObjectsRequest>, DB::S3::Client::DeleteObjects(DB::S3::ExtendedRequest<Aws::S3::Model::DeleteObjectsRequest> const&) const::$_11>(DB::S3::ExtendedRequest<Aws::S3::Model::DeleteObjectsRequest> const&, DB::S3::Client::DeleteObjects(DB::S3::ExtendedRequest<Aws::S3::Model::DeleteObjectsRequest> const&) const::$_11) const (this=0x7f9987305800, request=..., request_fn=...) at /build/src/IO/S3/Client.cpp:352
2023-05-01 18:22:35 #20 0x0000000029896a50 in DB::S3::Client::DeleteObjects (this=0x7f9987305800, request=...) at /build/src/IO/S3/Client.cpp:317
2023-05-01 18:22:35 #21 0x000000002a61cad3 in DB::S3ObjectStorage::removeObjectsImpl (this=0x7f9987330618, objects=..., if_exists=true) at /build/src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:353
2023-05-01 18:22:35 #22 0x000000002a61d322 in DB::S3ObjectStorage::removeObjectsIfExist (this=0x7f9987330618, objects=...) at /build/src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:379
2023-05-01 18:22:35 #23 0x000000002a5f5fc5 in DB::(anonymous namespace)::RemoveObjectStorageOperation::finalize (this=0x7f998566f060) at /build/src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:143
2023-05-01 18:22:35 #24 0x000000002a5f1719 in DB::DiskObjectStorageTransaction::commit (this=0x7f998573ba58) at /build/src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:808
2023-05-01 18:22:35 #25 0x000000002a59e9e4 in DB::DiskObjectStorage::removeSharedFile (this=0x7f9987330918, path=..., delete_metadata_only=false) at /build/src/Disks/ObjectStorages/DiskObjectStorage.cpp:275
2023-05-01 18:22:35 #26 0x000000002a5a8023 in DB::DiskObjectStorage::removeFile (this=0x7f9987330918, path=...) at /build/src/Disks/ObjectStorages/DiskObjectStorage.h:78
2023-05-01 18:22:35 #27 0x000000002cf72b8f in DB::MergeTreeMutationEntry::removeFile (this=0x7f989e7e6920) at /build/src/Storages/MergeTree/MergeTreeMutationEntry.cpp:104
2023-05-01 18:22:35 #28 0x000000002cf48863 in DB::StorageMergeTree::clearOldMutations (this=0x7f98e9728c40, truncate=false) at /build/src/Storages/StorageMergeTree.cpp:1400
2023-05-01 18:22:35 #29 0x000000002cf5e4aa in DB::StorageMergeTree::scheduleDataProcessingJob(DB::BackgroundJobsAssignee&)::$_4::operator()() const (this=0x7f998785fa40) at /build/src/Storages/StorageMergeTree.cpp:1324

2023-05-01 18:22:41 Thread 1 (Thread 0x7f9aaed60400 (LWP 191026)):
2023-05-01 18:22:41 #0  0x00007f9aaf0b4376 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
2023-05-01 18:22:41 #1  0x00000000327ebb19 in Poco::EventImpl::waitImpl (this=0x7f98bff50bb8) at /build/base/poco/Foundation/src/Event_POSIX.cpp:94
2023-05-01 18:22:41 #2  0x0000000022d7caf5 in Poco::Event::wait (this=0x7f98bff50bb8) at /build/base/poco/Foundation/include/Poco/Event.h:92
2023-05-01 18:22:41 #3  0x000000002ca734f8 in DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::removeTasksCorrespondingToStorage (this=0x7f9950443198, id=...) at /build/src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:161
2023-05-01 18:22:41 #4  0x000000002c9356a4 in DB::BackgroundJobsAssignee::finish (this=0x7f98e97291f0) at /build/src/Storages/MergeTree/BackgroundJobsAssignee.cpp:117
2023-05-01 18:22:41 #5  0x000000002cf3a342 in DB::StorageMergeTree::shutdown (this=0x7f98e9728c40) at /build/src/Storages/StorageMergeTree.cpp:194
```

```
2023.05.01 18:20:22.816089 [ 191022 ] {} <Debug> Application: Received termination signal.
...
2023.05.01 18:21:57.921177 [ 191316 ] {} <Trace> test_5.many_mutations (118a5f1d-1702-4bce-a9f9-9ffba7e40a0d): Removing mutation: mutation_19351.txt
# huge delay
2023.05.01 18:22:41.841491 [ 191277 ] {} <Information> KeeperDispatcher: Found dead session 4, will try to close it
2023.05.01 18:22:41.841665 [ 191277 ] {} <Information> KeeperDispatcher: Dead session close request pushed

2023.05.01 18:22:42.184140 [ 191316 ] {} <Trace> S3ObjectStorage: Objects with paths [test/hfa/itbrnmocnuydlqueeeyamgwjdxhuj] were removed from S3
2023.05.01 18:22:42.184558 [ 191316 ] {} <Trace> test_5.many_mutations (118a5f1d-1702-4bce-a9f9-9ffba7e40a0d): Removing mutation: mutation_19393.txt
```
